File: src/core/search/zobrist.ts
// src/core/search/zobrist.ts
import { BB, bits } from '../bitboards';
import { Position } from '../position';

// Zobrist: [pieceType(0..3)][square(0..31)] and sideToMove
// Types: 0=P1 man, 1=P1 king, 2=P2 man, 3=P2 king

const R = (seed: number) => () => {
  let t = (seed = (seed + 0x6D2B79F5) >>> 0);
  t = Math.imul(t ^ (t >>> 15), 1 | t);
  t ^= t + Math.imul(t ^ (t >>> 7), 61 | t);
  return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
};

const rand32 = (rng: () => number) => (rng() * 0x100000000) >>> 0;

const rng = R(0xC0FFEE);
export const Z_PIECE: number[][] = Array.from({ length: 4 }, () => Array(32).fill(0));
export const Z_SIDE: number = rand32(rng);

(function init() {
  for (let t = 0; t < 4; t++) {
    for (let s = 0; s < 32; s++) Z_PIECE[t][s] = rand32(rng);
  }
})();

export function hashPosition(p: Position): number {
  let h = 0 >>> 0;
  for (const i of bits(p.p1Men))   h ^= Z_PIECE[0][i];
  for (const i of bits(p.p1Kings)) h ^= Z_PIECE[1][i];
  for (const i of bits(p.p2Men))   h ^= Z_PIECE[2][i];
  for (const i of bits(p.p2Kings)) h ^= Z_PIECE[3][i];
  if (p.side === 1) h ^= Z_SIDE;
  return h >>> 0;
}

File: swift-native/Sources/MakhosCore/Zobrist.swift
private struct Mulberry32 {
    private var state: UInt32

    init(seed: UInt32) {
        state = seed
    }

    mutating func nextUInt32() -> UInt32 {
        state &+= 0x6D2B79F5
        var t = state
        t = (t ^ (t >> 15)) &* (1 | t)
        t ^= t &+ ((t ^ (t >> 7)) &* (61 | t))
        return t ^ (t >> 14)
    }
}

public enum Zobrist {
    public static let piece: [[UInt32]] = {
        var rng = Mulberry32(seed: 0x00C0FFEE)
        return (0..<4).map { _ in
            (0..<32).map { _ in rng.nextUInt32() }
        }
    }()

    public static let side: UInt32 = {
        var rng = Mulberry32(seed: 0x00C0FFEE)
        _ = (0..<(4 * 32)).map { _ in rng.nextUInt32() }
        return rng.nextUInt32()
    }()

    public static let verifyPiece: [[UInt32]] = {
        var rng = Mulberry32(seed: 0xDEADBEEF)
        return (0..<4).map { _ in
            (0..<32).map { _ in rng.nextUInt32() }
        }
    }()

    public static let verifySide: UInt32 = {
        var rng = Mulberry32(seed: 0xDEADBEEF)
        _ = (0..<(4 * 32)).map { _ in rng.nextUInt32() }
        return rng.nextUInt32()
    }()
}

public func hashPosition(_ position: Position) -> UInt32 {
    var hash: UInt32 = 0
    for index in bits(position.p1Men) { hash ^= Zobrist.piece[0][index] }
    for index in bits(position.p1Kings) { hash ^= Zobrist.piece[1][index] }
    for index in bits(position.p2Men) { hash ^= Zobrist.piece[2][index] }
    for index in bits(position.p2Kings) { hash ^= Zobrist.piece[3][index] }
    if position.side == .playerOne { hash ^= Zobrist.side }
    return hash
}

public func verifyHashPosition(_ position: Position) -> UInt32 {
    var hash: UInt32 = 0
    for index in bits(position.p1Men) { hash ^= Zobrist.verifyPiece[0][index] }
    for index in bits(position.p1Kings) { hash ^= Zobrist.verifyPiece[1][index] }
    for index in bits(position.p2Men) { hash ^= Zobrist.verifyPiece[2][index] }
    for index in bits(position.p2Kings) { hash ^= Zobrist.verifyPiece[3][index] }
    if position.side == .playerOne { hash ^= Zobrist.verifySide }
    return hash
}
